lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -2993,13 +3092,60 @@ AdjointValue PullbackCloner::Implementation::accumulateAdjointsDirect(
       }
       return makeAggregateAdjointValue(lhsVal->getType(), newElements);
     }
+    // x + (zeroBase, index, eltToAdd) => (x.0, ..., x.index + eltToAdd, ...,
```

Comment:
why can't you simply call `accumulateAdjoints` on `x` and `base`, and then wrap the result with `AddElement` again? This way you'd evaluate everything symbolically and only materialize the real value (emit addition) when necessary. It will also collapse zeros on fly.

Reply:
Fixed in the latest revision.